File: leetcode/LeetCode_697_degree.cpp
class Solution {
public:
    int findShortestSubArray(vector<int>& nums) {
        int len = nums.size();
        int max_num = *max_element(nums.begin(),nums.end()) + 1;
        vector<int> max_(max_num, 0);
        vector<int> min_(max_num, 0x7FFFFFFF);
        vector<int> cnt_(max_num,0);


        for(int i = 0;i < len;i++)
        {
            max_[nums[i]] = max(max_[nums[i]], i);
            min_[nums[i]] = min(min_[nums[i]], i);
            cnt_[nums[i]]++;
        }

        int degree = *max_element(cnt_.begin(),cnt_.end());
        int ans = len;
        for(int i = 0;i < len;i++)
        {
            if(cnt_[nums[i]] == degree)
            {
                ans = min(ans, max_[nums[i]] - min_[nums[i]] + 1);
            }
        }
        return ans;
    }
};

File: leetcode_archived/LeetCode_697_degree.py
class Solution:
    def findShortestSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        max_ = {}
        min_ = {}
        cnt_ = collections.defaultdict(int)


        for key, val in enumerate(nums):
            max_[val] = max(max_.get(val,-1), key)
            min_[val] = min(min_.get(val,0x7FFFFFFF),key)
            cnt_[val] += 1

        degree = max(cnt_.values())
        ans = len(nums)

        for val in nums:
            if cnt_[val] == degree:
                ans = min(ans, max_[val] - min_[val] +1)
        return ans
